Issue #1739: add view and buildContent methods on Entity Controller

herbdool opened the pull request:
Fixes https://github.com/backdrop/backdrop-issues/issues/1739

docwilmot commented:
This is great to see!
No `view()` method on `Entity` class though?

herbdool commented:
@docwilmot I tried to keep it in line with `entity_plus` by putting the methods on the controller.

The PR requires these methods since they're in the EntityControllerInterface, but if we wanted to make it more backwards compatible they should probably optional. Not sure which is best.

docwilmot commented:
We'd keep the method in the controller but also call get controller from the entity view method. So both available. We need to be able to call a simple node->view(). For example.

herbdool commented:
@docwilmot gotcha. I've added the view method on Entity as well.

All tests passing.

docwilmot commented:
Don't mean to be greedy, but I suspect we need an `Entity::buildContent()` too.

herbdool commented:
@docwilmot @argiepiano one thing I noticed on my local install is that each entity type still seems to call `template_preprocess_entity`. I assumed that, say, `view` on the Node controller would override `view` on the Default Entity controller. But perhaps `entity_theme` is just a bit too greedy here. https://github.com/backdrop/backdrop/blob/2d1373bcfe777652ad2dfa78a19dcc7cd68576fe/core/modules/entity/entity.module#L49. I had more or less copied it from `entity_plus` so it could use some introspection.

docwilmot commented:
Not sure what you mean. If they all use `entity_theme()` then they'd all invoke `template_preprocess_entity()` right?

argiepiano commented:
> @docwilmot @argiepiano one thing I noticed on my local install is that each entity type still seems to call `template_preprocess_entity`. I assumed that, say, `view` on the Node controller would override `view` on the Default Entity controller. But perhaps `entity_theme` is just a bit too greedy here.

@herbdool  I also assumed that, since you are overriding `view` in the node controller, you are also rewriting `#theme` there, which would result on the node preprocess functions to be called instead of the entity preprocess. I'll take a closer look tonight.

argiepiano commented:
@herbdool after some testing I can confirm that the correct preprocess hooks are invoked (the ones set by each entity's controller, i. e. template_preprocess_node for nodes, etc). I tested nodes, users, files, taxonomies and comments. So, all looks great. I also checked that the view and buildContent methods are being invoked correctly in all of the new controllers.  

I also checked to see if template_preprocess_entity was called at all when viewing any of the above entities. It's not. 

If you are still getting calls to template_preprocess_entity for some reason, maybe we are missing something and we need to do some more digging.

BTW I also did some tests with Entity Plus enabled, and I saw no issues.

herbdool commented:
@argiepiano I'm not sure if it's calling `template_preprocess_entity` but when in git I switch from this branch to the 1.x branch, I get this error:

```
Warning: include_once(/app/core/modules/entity/entity.theme.inc): failed to open stream: No such file or directory in include_once() (line 1103 of /app/core/includes/theme.inc).
Warning: include_once(): Failed opening '/app/core/modules/entity/entity.theme.inc' for inclusion (include_path='.:/usr/local/lib/php') in include_once() (line 1103 of /app/core/includes/theme.inc).
```

which makes me wonder why it's loading that file when it should be using the node and comment functions in this case (on a post with comments).

herbdool commented:
@argiepiano the other clue is that this PR messes up the styling of the comments. No nice box around the comment until I switch back to 1.x and clear cache.

docwilmot commented:
OOps. Closed accidentally.

herbdool commented:
@argiepiano the PR is updated. I think it's still calling `template_preprocess_entity`.

argiepiano commented:
@herbdool I'm testing and working on this right now. I do see `template_preprocess_entity` being called for `comment` entities only. This is unlike all other entities (node, file, user). All other entities are correctly  calling their preprocessors. This is a stubborn bug I'm determined to find. 

The reason why you are not seeing the nice box around the comments is because `template_preprocess` is adding the css class `entity` instead of `comment` to the class array. That will be solved once I solve the other issue.

I'll report once I have news - back to beating the crap out of dpm().

BTW: there is an unrelated separate issue with the PR that I will investigate later: if you download Backdrop 1.x to your local, then apply the patch for this PR, and then try to do a clean install (using `/install.php` with an empty database) you get WSOD. If, instead, you do a clean install BEFORE patching, and later apply the patch, then everything works fine. I think it has to do with the order in which hook_entity_info is run when you apply the patch before doing the install.  I'll try to tackle that after this other crazy preprocessing thing

argiepiano commented:
@herbdool , after extensive testing and tracing this is what I'm seeing. I'm going to use `comment` as an example, but this is happening with all entities.  Bear with me:

- `template_preprocess_entity` is called ONLY if you use a tpl.php file inside your theme's templates folder. If you stick with the tpl files provided in the core modules (that is, core/modules/comment/templates/comment.tpl.php), the `template_preprocess_entity` is not invoked at all
- This is not too problematic, because in fact `template_preprocess_comment` IS invoked right after `template_proprocess_entity`
- However, when this happens,  the css class `comment`  is not applied to the output (instead `entity` css class is applied), resulting in the loss of the nice square frame around the comments that you noticed in an earlier comment.

This missing `comment` css class can be added easily inside `template_preprocess_comment`. Tomorrow, I'll create some suggestions  within your code to do just that for all entities. 

I spent some time figuring out the reason why `template_preprocess_entity` is being invoked at all. When Backdrop finds a tpl file within a templates folder of a theme, it adds an element to the theme registry array for that particular entity, like so: `'base hook' => 'entity'`. This base hook element forces the call to `template_preprocess_entity`. The addition of `base hook` is happening because entity_theme() declares a `pattern` key that includes all known entities to Backdrop (see line 67 of your PR of `entity.module`). If we eliminated that `pattern` from the array, Backdrop would stop invoking `template_preprocess_entity`, BUT it somehow loses the ability to find  derivative tpl.php files inside the templates folder of your theme. I need to investigate further to see if this is a bug.

Anyway, bottom line: at this point `template_preprocess_entity` will be called before `template_preprocess_comment` if you place a tpl.php template in the templates folder of your theme. BUT this should not present any serious issues - so far the only problem I see is that the output loses the css class "comment", which can be easily added when preprocessing the comment. 

The other bug I'm encountering is what I described above in another comment - the WSOD screen when you patch Backdrop 1.x with your PR before doing a clean install. I'll take a look at that in the next couple of days.

Thanks for all the work!

argiepiano commented on core/modules/entity/entity.controller.inc:
```diff
@@ -592,49 +599,43 @@ class DefaultEntityController implements EntityControllerInterface {
+      );
+      // Allow modules to modify the structured entity.
+      backdrop_alter(array($this->entityType . '_view', 'entity_view'), $build, $this->entityType);
+      $view[$this->entityType][$entity->id()] = $build;
```
The abstract class `Entity` currently does not currently provide a `id()` method, even though its definition it's included in the interface `EntityInterface`.  We could create this method.

herbdool replied:
It's not needed as part of this PR - there are other parts of code that are calling `$entity->id()` too.

argiepiano commented:
I just left a comment in the PR code regarding the absence of method `Entity::id()`, which is invoked in `EntityDefaultController::view()`. 

This brings up a bigger question: currently core does not provide a "true" `Entity` class (it provides an abstract class called `Entity` - which cannot be instantiated in PHP). Would it be a good idea to provide an "out of the box" ready to use `Entity` class so that people creating custom entities with core functionality are not forced to write their own? (side note: Entity Plus provides such a class).

argiepiano commented on core/modules/entity/entity.theme.inc:
```diff
@@ -0,0 +1,56 @@
+  list(, , $bundle) = entity_extract_ids($entity_type, $entity);
+
+  // Gather css classes.
+  $variables['classes_array'][] = backdrop_html_class('entity-' . $entity_type);
```
These two lines create `$variables['classes_array']`. However, `entity.tpl.php` uses $classes rather than $classes_array for the css class markup.

I suggest adding these css classes to `$variables['classes']` and getting rid of `$variables['classes_array']` in these two lines.

This will assure that `entity.tpl.php` will also receive a css class with [entityType]-[bundle] (which currently it doesn't since it never uses $classes_array).

docwilmot commented:
> This brings up a bigger question: currently core does not provide a "true" Entity class (it provides an abstract class called Entity - which cannot be instantiated in PHP)..

There was a reason for that: https://github.com/backdrop/backdrop-issues/issues/2558#issuecomment-295862718

argiepiano commented on core/modules/entity/entity.module:
```diff
@@ -43,6 +43,33 @@ function entity_flush_caches() {
+  $types = array();
+  foreach (entity_get_info() as $type => $info) {
+    if (isset($info['controller class']) && in_array('EntityControllerInterface', class_implements($info['controller class']))) {
+      $types[] = $type;
```
I suggest: 
```php
      if (!in_array($type, array('node', 'comment', 'file', 'taxonomy_term', 'user'))) {
        $types[] = $type;
      }
``` 
This will leave all 5 core entities out of the pattern, resulting in `template_preprocess_entity` not being called for those, since they provide their own full preprocess functions

argiepiano commented on core/modules/entity/entity.module:
```diff
@@ -43,6 +43,33 @@ function entity_flush_caches() {
+  // This has to match the template suggestions provided in
+  // template_preprocess_entity().
+  $types = array();
+  foreach (entity_get_info() as $type => $info) {
```
The following lines of code are causing a PHP Fatal Error when doing a "clean" install of a patched Backdrop with this PR. While I don't have an exact cause, I believe the issue is that this line is calling `entity_get_info` before all the necessary inc files have been loaded (this is happening during the first stages of the install cycle).

I'm not sure why we haven't experienced this Fatal Error with the Tugboat test sites. Perhaps Tugboat doesn't do a clean install from scratch of Backdrop? Or perhaps it applies these patches after installing? (which doesn't cause the error).  

I'm haven't come up with a solution for this yet. These lines are needed so that we have access to the appropriate template files for custom entities "out of the box" . Perhaps this `pattern` can be inserted in  an implementation of `hook_theme_registry_alter`? I haven't tested this approach yet. Will report after I do.

[EDIT] I can also confirm that the install works fine if these lines are commented out after patching and before installing.

docwilmot replied:
Do you mind explaining how you make the patched Backdrop?

argiepiano replied:
No problem. I download Backdrop 1.x from the repository, then download this PR as a patch here: https://patch-diff.githubusercontent.com/raw/backdrop/backdrop/pull/3848.patch

Then I apply the patch BEFORE installing Backdrop running a `patch` command in shell. Then install the site normally from the browser

docwilmot replied:
Youre right, fatal error confirmed. On loading `install.core.php` Backdrop loads modules User, Entity and System. Now that Entity calls `entity_get_info()` immediately, this calls `user_entity_info()` which calls `db_table_exists()` which doesnt yet exist because the `database.inc` file isnt yet loaded.

I can solve this by wrapping that code in 
```
  global $databases;
  if (!empty($databases)) {
    foreach (entity_get_info() as $type => $info) {
    ... etc
```
Or we could probably check bootstrapping level.

herbdool replied:
I guess the only reason we don't have the same problem with entity_plus is that it can't be installed at the same time as core.

Is there another way of getting the "controller class" without using entity_get_info()?

If an `entity` in core is just abstract maybe we should get rid of `entity_theme` altogether. The code was copied from entity_plus where there is a "real" generic entity class. I think @argiepiano you mentioned something about that breaking something else, but I may dig into that anyway.

herbdool commented:
I've taken a radical approach of ripping out `entity_theme`, `entity.theme.inc` and `entity.tpl.php`. I think the better approach here is to *not* provide some kind of "fallback" for other entities. Entity Plus has a basic entity, core doesn't need it and it just causes more issues than it solves.

I bet most contrib modules with entities already provide their own theme functions and templates. I checked `flag`, `paragraphs`.

argiepiano commented:
> I've taken a radical approach of ripping out `entity_theme`, `entity.theme.inc` and `entity.tpl.php`. I think the better approach here is to _not_ provide some kind of "fallback" for other entities. Entity Plus has a basic entity, core doesn't need it and it just causes more issues than it solves.

@herbdool I'm not sure this approach is wise in the long run, but I need to give this some extra thought. You have added `view` and `buildContent` to `DefaultEntityController`. In particular, `view` returns a renderable array. AFAIK all renderable arrays need to include a `'#theme'` key - otherwise the render engine will not know how to theme the array (I think). Have you tested this? 

At this point, also notice that you have kept line 628 in `entity.controller.inc` that reads:

```php
'#theme' => 'entity',
```

I think that will throw an error, since now there is no definition of that theme anymore. 

In my view, it's better not to take a radical approach, but rather  to follow @docwilmot suggestion of wrapping those lines in `entity_theme` as he suggested.

herbdool commented:
@argiepiano all that fiddling smells wrong to me (we could always add stuff back in if I'm wrong). Since we're talking about an abstract class, we don't really want it to have concrete effects all on its own. We could even remove view/buildContent from the DefaultEntityController entirely but right now the PR has them in the Interface so they have to be there and we want all entities to define them anyway. So they're in the abstract class but I stripped them down so they just return an empty array.

I think it would be good to get @quicksketch or @hosef to provide some feedback if we're on the right track.

docwilmot commented:
I dont agree with this direction either.

herbdool commented:
I appreciate the feedback.

I can see how the DefaultEntityController needs useful view method that can be used by custom entities (the comment on the controller mentions how entities can use it directly) so what I’ve got in the PR is incomplete. However, I don’t like the messy workarounds suggested above to get around the bugs.

I don’t think checking for the $database global will help insofar as it would probably mean entity_theme would never return a pattern, no? Or if it’s only happening on install it still would mean the registry gets cached so it would require a cache clear I think.

And I don’t like hardcoding exceptions to the types either. What if a new entity is added to core? Who will remember to go back to this code to update it? 

Messy.

argiepiano commented:
I'm new to this community and don't have the insight of those of you who have been immersed in the philosophy of Backdrop for many years. But let me see if I'm getting this right:

1. The goal of creating `entity_view` in core and adding `view` and `buildContent` to `DefaultEntityController` is to provide a higher level of abstraction that will allow modules like entityreference to provide themed views of the entities they reference, regardless of what and where that entity might be (core, custom, etc)
2. `entity_theme` is there to provide a generic theme (`entity`) that allows the theming of custom entities "out of the box," without the need for those custom entities to implement their own theme, template and preprocessing function. 
3. All core entities already provide their own theming, templates and preprocessing, so `entity_theme` (and the `pattern`) is not really needed in those cases
4. The  use of `pattern` in  `entity_theme` is there to assure that custom entities that do not define their own themes, templates and preprocessing functions can use template files such as `CUSTOM_ENTITY.tpl.php` out of the box without defining specific `theme_hook_suggestions`. 
5. But you saw how confusing that `pattern` approach can be -- see comments in this issue (above), in a few Drupal 7 Paragraphs threads, and in recent issues in Backdrop Paragraphs:  `template_preprocess_entity_plus` was STILL being called despite the fact that the theme engine used `paragraphs_item.tpl.php` . And above, you saw how  `template_preprocess_entity` was still being called when you placed `comments.tpl.php` in a theme folder. This is so because `entity` (or `entity_plus` in the case of Paragraphs) was being defined as their `base hook` by the use of that `pattern` 

It seems to me that the easiest solution to this messy stuff is to simply expect custom entities to provide their own theming and to do a light override of DefaultEntityController and buildContent to specify the #theme to be used in the render array output. We saw recently in Paragraphs how simple extending  DefaultEntityPlusController and overriding `buildContent` with [just three lines](https://github.com/backdrop-contrib/paragraphs/pull/114/commits/e0475255a2062ecfb5088243dd88c46d68bfda1a) accomplished what was needed to assure that `template_preprocess_paragraphs_item` was being invoked, and `paragraphs_item.tpl.php` was used for theming.

So, from my point of view, I would keep `entity_theme` as a fallback theme, but without the controversial `pattern` stuff. This will force custom entities that do not define their own theme/preprocessing/templates or extend DefaultEntityController to use `template_preprocess_entity` and the generic template file `entity.tpl.php`.

My 2 cents. Again, I defer to those of you who have more experience with Backdrop. And I'm sure there are other solutions.

herbdool commented:
> I would keep entity_theme as a fallback theme, but without the controversial pattern stuff. This will force custom entities that do not define their own theme/preprocessing/templates or extend DefaultEntityController to use template_preprocess_entity and the generic template file entity.tpl.php.

@argiepiano Sounds like it's worth a try. I am also still learning a lot about the innards. I've not dug into all of it all this time I've been using Drupal 7 and Backdrop.

I'm going to test this tomorrow.

You said previously:

>  If we eliminated that pattern from the array, Backdrop would stop invoking template_preprocess_entity, BUT it somehow loses the ability to find derivative tpl.php files inside the templates folder of your theme. I need to investigate further to see if this is a bug.

Hopefully this won't be an issue.

argiepiano commented:
> You said previously:
> 
> > If we eliminated that pattern from the array, Backdrop would stop invoking template_preprocess_entity, BUT it somehow loses the ability to find derivative tpl.php files inside the templates folder of your theme. I need to investigate further to see if this is a bug.
> 
> Hopefully this won't be an issue.

@herbdool when I did testing a few days ago, this was an issue. But if we go the route of providing a generic entity.tpl.php we could modify template_preprocess_entity to provide generic derivative theme_hook_suggestions for entity.tpl.php

One thing we need is a custom entity with which to test your commits (testing with core entities is easy, but we are in the dark with custom entities). I'm forking the `entity_example` provided as part of Examples for Developers contrib module, just for testing. I'll put it in my github and let you know once it's there.

herbdool commented:
@argiepiano I've added the theme and template back but with no pattern. I changed the `theme_hook_suggestions` to start with `entity__` because I figure that it should follow the pattern of `entity{--ENTITY_TYPE...}.tpl.php`. Then if someone wants a more specific template they can add their own theme function.

At first I thought the suggestions weren't working but turns out they were. As you recommended I used `entity_example` for testing this. Turns out the module mostly worked but it was missing some things from its entity class to work properly. I'll submit a PR to that project too.

argiepiano commented:
> At first I thought the suggestions weren't working but turns out they were. As you recommended I used `entity_example` for testing this. Turns out the module mostly worked but it was missing some things from its entity class to work properly. I'll submit a PR to that project too.

That's great, @herbdool . I'm doing a quick adaptation of example_entity in two versions, one with and without without the controller override and theme definition. (I won't submit a PR for that project - just doing a quick and dirty adaptation for testing with your PR). I think I can have that custom entity example done by 5pm US Mountain Time today (busy day today). I'll share the repository with you so that you can also test.   

> I changed the theme_hook_suggestions to start with entity__ because I figure that it should follow the pattern of entity{--ENTITY_TYPE...}.tpl.php. Then if someone wants a more specific template they can add their own theme function.

That makes sense.

argiepiano commented:
@herbdool , I have baked an example entity module based on `entity_example` called `panettone` (yes, it's the holiday season). 

https://github.com/argiepiano/panettone

It has two branches. The main branch uses the default `EntityDatabaseStorageController` and therefore it uses the `entity` theme, preprocessing and templates . 

The branch `own_controller` defines a theme, preprocessor and templates, and lightly extends  `EntityDatabaseStorageController` to correctly theme the output. 

Also notice that `entity_example` needed quite a few modifications to make it work properly with the current default controller.  I also needed to fill some of the methods in the entity class provided by `entity_example`,  such as uri() and id().  

I've done quite a bit of testing. Everything is working **GREAT**!   Templates are working well when placed both in the modules and the themes folders, and preprocessing functions are being called correctly.

I think we have a winner version with your latest PR. 🎉 🎉 🎉

herbdool commented:
@argiepiano thank you! Nice to have it all working. Now my mouth is watering.

docwilmot commented:
Is there an issue label for mouth-watering?

docwilmot commented:
@argiepiano can you do a code review please?

argiepiano commented:
> @argiepiano can you do a code review please?

I'll be glad to help. Besides reviewing the new code and doing some testing, is there anything else involved in reviewing code? And once I complete that, how do I report it? In the original issue or here? Do I change any of the labels?

herbdool commented:
@argiepiano once you've reviewed you update the labels on the related issue and comment there. If you think it's ready then add the label to "ready to be committed" and if you tested "works for me". That's it!

Then a core commiter will review and merge if it's all good.

If it's not ready you add the label to "needs work".

argiepiano commented:
I'm in the middle of reviewing the code - quite a complex task given the large scope of the changes. I'm focusing on Comment for now.

One thing I noticed is that `comment_view_multiple` may need to be rewritten too. I'm encountering an error when manually invoking `comment_view_multiple` that was not there in the old `comment_view_multiple`.  I haven't dug enough to find the cause, but I believe it has to probably do with hook invocations.  

**How to reproduce**
(I'm working locally with a clean install of Backdrop, which I've patched with the PR provided here). 
- Create a post and add comment
- With devel try the following:

```php
$c = comment_load(5); // Replace with the correct cid
$n = node_load(3); // Replace with the correct nid
dpm(comment_view_multiple(array($c), $n));
```
I'm consistently getting:
```Notice: Undefined offset: 5 in _field_invoke_multiple() (line 293 of /Users/alejandrocremaschi/Sites/localhost/bdentity/core/modules/field/field.attach.inc).```

(where the offset coincides with the cid).

This is not triggered if I do exactly the same on a clean install without the patch.

I will investigate a bit further later.

herbdool commented:
Thanks @argiepiano. I took a look at `comment_view_multiple` and `comment_build_content` and noticed a stray parameter: `, array('node' => $node)` from my previous approach. Let's see if it passes your manual test now.

Perhaps this is missing from the tests.

herbdool commented:
Though they might not be connected - I didn't see a direct path from `comment_view_multiple` to `comment_build_content`.

argiepiano commented:
@herbdool , I found the issue - it wasn't a bug. I was passing a regular array of comments to `comment_view_multiple`, but that array needed to be keyed by cid (that's the way `comment_load_multiple` returns it). 

I see that you found that astray argument (which was unrelated to this issue). Great. That was one of the comments I was about to submit 😉

argiepiano commented on core/modules/comment/comment.entity.inc:
```diff
@@ -435,4 +435,102 @@ class CommentStorageController extends EntityDatabaseStorageController {
+   */
+  public function buildContent(EntityInterface $comment, $view_mode = 'full', $langcode = NULL, $context = array()) {
+    if (!isset($langcode)) {
+      $langcode = $GLOBALS['language_content']->langcode;
```
All these references to `$GLOBALS['language_content']->langcode` can be replaced by:

```php
global $language_content;
...
$langcode = $language_content->langcode;
```

They happen in all new buildContent methods. I believe it's a cleaner way, and it's used in many other places in Backdrop.

herbdool replied:
I've changed them. Actually still with a conditional because the `$langcode` could still be already set.

I noticed a couple of the previous format though but this is better.

argiepiano commented on core/modules/entity/templates/entity.tpl.php:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @see template_preprocess()
+ * @see template_preprocess_entity()
+ * @see template_process()
```
template_process() was removed from Backdrop a while ago. See [this notice](https://docs.backdropcms.org/change-records/removed-theme-process-layer).

Please check other tpl files for these issues.

argiepiano commented on core/modules/entity/templates/entity.tpl.php:
```diff
@@ -0,0 +1,47 @@
+ * @file
+ * Default theme implementation for entities.
+ *
+ * Available variables:
```
IT'd be a good idea to add the other variables created in the preprocess step:  $entity_type and $view_mode

herbdool replied:
I think I've addressed most of this.

One thing: I had added `$title_attributes` and `$content_attributes` in the template but I don't see them in the preprocessor. Are they actually present or did I just partly copy/paste from elsewhere?

argiepiano replied:
'title_attributes' is added by `system_preprocess_page_components` and passed through the variables to other preprocessors. 'content_attributes' gets an empty default value in `template_preprocess`. They are real variables, but I'm not sure if they should be listed in the php documentation in tpl

argiepiano commented on core/modules/entity/templates/entity.tpl.php:
```diff
@@ -0,0 +1,42 @@
+ * - $title: The (sanitized) entity label.
+ * - $url: Direct url of the current entity if specified.
+ * - $page: Flag for the full page state.
+ * - Array of html class attribute values. By default the following classes are
```
Add the name of the variable here - $classes: An array of html....
It may be a good idea to also add $attributes ? (this is added by template_preprocess())

argiepiano commented on core/modules/file/file.entity.inc:
```diff
@@ -241,4 +241,94 @@ class FileStorageController extends EntityDatabaseStorageController {
+      unset($file->content['file']['#language']);
+    }
+    else {
+      unset($file->content['file']);
```
This line (and the whole else statement) doesn't do anything. We get here only if $file->content['file'] is not set.

herbdool replied:
All of this was moved over from `file_build_content` so that comment has been around for awhile. There's actually another issue to fix this https://github.com/backdrop/backdrop-issues/issues/5370. If it's merged we'll need to fix it here too.

argiepiano replied:
OK

argiepiano commented on core/modules/file/file.entity.inc:
```diff
@@ -241,4 +241,94 @@ class FileStorageController extends EntityDatabaseStorageController {
+    entity_prepare_view('file', array($file->fid => $file), $langcode);
+
+    // Build the actual file display.
+    // @todo Figure out how to clean this crap up.
```
😄 agreed!

argiepiano commented on core/modules/taxonomy/taxonomy.entity.inc:
```diff
@@ -297,4 +297,80 @@ class TaxonomyTermController extends EntityDatabaseStorageController {
+    // Add the term description if the term has one and it is visible.
+    $type = 'taxonomy_term';
+    $entity_ids = entity_extract_ids($type, $term);
+    $settings = field_view_mode_settings($type, $entity_ids[2]);
```
This variable is never used in this function.

herbdool replied:
I'm leaning towards creating follow-ups for this since all of this was simply moved over from `taxonomy_build_content`.

argiepiano commented on core/modules/user/user.entity.inc:
```diff
@@ -495,4 +495,61 @@ class UserStorageController extends EntityDatabaseStorageController {
+  /**
+   * Implements EntityControllerInterface::buildContent().
+   */
+  public function buildContent(EntityInterface $account, $view_mode = 'full', $langcode = NULL, $context = array()) {
```
$account is an instance of User. The User class extends Entity (which implements EntityInterface), but the User class also implements UserInterface, which is more specialized than EntityInterface. 

I'm thinking it may be best to declare the argument as UserInterface to be clearer.

argiepiano replied:
Actually I'm wrong here. The argument declaration has to match the method it's extending

argiepiano commented on core/modules/user/user.module:
```diff
@@ -2270,27 +2270,7 @@ function user_view_page($account) {
  *   An array as expected by backdrop_render().
  */
 function user_view($account, $view_mode = 'full', $langcode = NULL) {
```
To keep the same pattern as in other view functions, we should user (User $account, ...) here (declare the type of argument)

herbdool replied:
Let's make this a follow-up as well.

argiepiano commented on core/modules/user/user.module:
```diff
@@ -2307,28 +2287,7 @@ function user_view($account, $view_mode = 'full', $langcode = NULL) {
  *   content language of the current request.
  */
 function user_build_content($account, $view_mode = 'full', $langcode = NULL) {
```
(User $account, ...)

herbdool replied:
Another follow-up

argiepiano commented:
I will try some focused testing tomorrow. For now, these (above) are things I've found just looking at the code.

Question:  I noticed that all your new `view` functions return only the build array, instead of the whole array, which is structured as `$view[ENTITY_TYPE][ENTITY_ID]`.  You do this with all of them

```php
$build = $view['user'][$account->uid];
return $build;
```


However Backdrop's render engine is smart enough to find the necessary data inside the full array. In my testing I've noticed that the "digging" up the build array is not needed. You can just return the whole array, and backdrop_render will do just fine. Check that out in entity_plus for example, where `entity_plus_view` returns the whole array. 

I'm just thinking it's cleaner, and it saves you 5 lines of code 😄

argiepiano commented:
I see an issue with the new function `entity_view`: it does not match the argument pattern used by D7's Entity API's `entity_view`.

D7's Entity API `entity_view`:
```php
function entity_view($entity_type, $entities, $view_mode = 'full', $langcode = NULL, $page = NULL) {}
```

New Backdrop's `entity_view`:
```php
function entity_view(EntityInterface $entity, $view_mode = 'full', $langcode = NULL) {}
```

In my view, one of the goals of creating all this new code was to provide an easy bridge between contrib modules that used Drupal 7 Entity API, so that they would not have to rely on Entity Plus.  For example entityreference, which makes use of the generic `entity_view` to display entities of any type, including custom entities. 

I realize you created the new `entity_view` to match core entities' view functions like `node_view`. However, this argument pattern will defeat the original purpose. 

My suggestion is to match D7's Entity API `entity_view` definition, thus facilitating the porting of dozens of contrib modules that  relied on Entity API. Given that `entity_view` did not exist in Backdrop, there is no potential backward compatibility issue. And of course we should not touch the arguments of `node_view`, `comment_view` etc since those  already existed.

argiepiano commented:
@herbdool I've done testing with core and custom entities and everything is looking good. I'm ready to mark this as "works for me", but am wondering about two aspects:
- My comment about the parameter pattern for `entity_view` above. Perhaps @docwilmot can also weigh in about my suggestion there
- The need to create automated tests, especially for `entity_view`, and perhaps for buildContent and view methods

AND thank you for all the work with this! 🙏🏽

docwilmot commented:
I agree we should try and match D7s pattern for `entity_view()`. I think rendering of all of the entities in core are well tested by existing tests though, so just a quick custom entity to make sure `theme_entity()` and the tpl is working as expected.

And also to keep in step with D7, we should also add `Entity::buildContent()`.

herbdool commented:
> Backdrop's render engine is smart enough to find the necessary data inside the full array. In my testing I've noticed that the "digging" up the build array is not needed.

@argiepiano alright I'll simplify this.

> My suggestion is to match D7's Entity API entity_view definition...

There's a difference between the classes' methods and this function though. I think it's good to maintain compatibility of the `view` and `buildContent` methods since the `entity_plus` class extends the core controller and would break if they're different. But `entity_view` can stand on its own, and it wouldn't be hard to fix any contrib modules to use the new format. I believe it's better to maintain a uniform "signature" for the `*_view()` functions (aside from that damn comment_view).

herbdool commented:
@argiepiano regarding "Backdrop's render engine is smart enough to find the necessary data inside the full array". Actually it looks like simplifying this causes fatal errors in the tests. I forgot that I had already gone through that.

docwilmot commented:
> There's a difference between the classes' methods and this function though. I think it's good to maintain compatibility of the view and buildContent methods since the entity_plus class extends the core controller and would break if they're different. But entity_view can stand on its own, and it wouldn't be hard to fix any contrib modules to use the new format. I believe it's better to maintain a uniform "signature" for the *_view() functions (aside from that damn comment_view).

A reasonable compromise then would be a new `entity_view_multiple()`. We ended up doing a similar thing for `entity_load()` and `entity_load_multiple()`  time ago.

herbdool commented:
@docwilmot good idea. I've added `entity_view_multiple()` that matches `entity_plus_view()`.

docwilmot commented on core/modules/entity/entity.module:
```diff
@@ -256,6 +256,39 @@ function entity_view(EntityInterface $entity, $view_mode = 'full', $langcode = N
+ *   there is no information on how to view an entity, FALSE is returned.
+ */
+function entity_view_multiple($entities, $view_mode = 'full', $langcode = NULL, $page = NULL) {
+  $entity_type = $entity->entityType();
```
Note `$entity` is undefined. 
Ideally should be `entity_view_multiple($entity_type, $entities, $view_mode = 'full', $langcode = NULL, $page = NULL)` or we'd need to reset to get the first entity, and/or `foreach` to be sure its not a mix of entities, which is messier I think.
edited.

herbdool replied:
oops

herbdool commented:
One thing that should be a follow-up is handling extra fields (pseudo fields). I noticed that this won't render them and I think each core entity has handled it in their own way. Entity Plus has a whole controller `EntityDefaultExtraFieldsController` for adding this functionality.

herbdool commented:
Either @docwilmot @argiepiano able to create a test for theme_entity and share it with me to add it to this PR?

docwilmot commented:
OK, can do

docwilmot commented:
Can one PR to a PR?

docwilmot commented on core/modules/entity/entity.theme.inc:
```diff
@@ -0,0 +1,60 @@
+  $variables['title'] = check_plain(entity_label($entity_type, $entity));
+
+  $uri = entity_uri($entity_type, $entity);
+  $variables['url'] = $uri && !empty($uri['path']) ? url($uri['path'], $uri['options']) : FALSE;
```
The `Entity::uri()` might not always return an `options` key. Before this line: `$uri += array('options' => array());`

herbdool replied:
Fixed

docwilmot commented:
Cant figure out how to PR without creating a whole new PR so until then:
```
In entityAPITestCase

  /**
   * Tests basic CRUD functionality of the Entity API.
   */
  function testEntityView() {
    $user1 = $this->backdropCreateUser();

    // Create a test entity.
    $entity = entity_create('entity_test', array('name' => 'test', 'uid' => $user1->uid));
    $entity->field_test_text[LANGUAGE_NONE][0]['value'] = 'Some random text';
    $entity->save();

    $this->backdropGet('test/' . $entity->id());
    $this->assertRaw('<h1 class="page-title">Entity test</h1>');
    $this->assertText('Some random text', 'Found the field text on the rendered entity.');
  }
```
and 

```
IN entity_test.module

function entity_test_menu() {
  $items['test/%entity_test'] = array(
  'title' => 'Entity test',
  'page callback' => 'entity_view',
  'page arguments' => array(1),
  'access callback' => TRUE,
  'type' => MENU_CALLBACK,
  );
  return $items;
}
```

I feel that the other core entities test other template and theme configurations. We can discuss if we need to test anything more than this.

herbdool commented:
@docwilmot thanks! I'll add it in

argiepiano commented:
Thank you both for all this work. You guys were on fire today! I've been traveling all day and have not had a chance to participate in the discussion. Great idea about adding `entity_view_multiple()`. Is that playing nice with backdrop_render? 

Regarding the extra (pseudo) fields rendering - I have often taken care of that by overriding buildContent and returning the extra fields as part of the $build.  

I have occasionally used `EntityDefaultExtraFieldsController` (you have to explicitly declare its use in the hook_entity_info with the key `'extra fields controller class' =>`). I've found it to be a bit of a "brute force" approach. For example, even if you choose not include some of your pseudo fields in your `hook_field_extra_fields()` declaration, that controller will still pick those properties and render them at the top of your display - very annoying. You have to declare the pseudo fields, and then hide them in the manage display settings for them not to appear. Anyway, something to discuss in a separate issue.

argiepiano commented:
> Can one PR to a PR?

The way I've seen it done is by forking the contributors fork of backdrop, and then doing a PR to the contributors fork, not to the original repository. If the contributors merges that PR into their fork, that creates a commit that then gets added to the PR of the original repository. Or something like that...

argiepiano commented:
> Great idea about adding `entity_view_multiple()`. Is that playing nice with backdrop_render?

I've done some testing (including theming and rendering) with core entities and custom entities and `entity_view_multiple()` works as expected.

herbdool commented:
I've added @docwilmot's test and everything is passing.

herbdool commented:
Thank you for the explanation about extra fields @argiepiano

docwilmot commented:
@argiepiano you've done several code reviews and I believe you seem happy with the code. Can you mark the issue RTBC? Note that we tend to have two levels of review: "works for me" means youve tested it, whereas RTBC means its also passed your code review. But I noticed you marked "works for me" after what seemed like a thorough code review.

quicksketch commented on core/modules/comment/comment.module:
```diff
@@ -935,55 +935,8 @@ function comment_prepare_thread(&$comments) {
-  backdrop_alter(array('comment_view', 'entity_view'), $build, $type);
-
-  return $build;
+  $view = entity_get_controller('comment')->view(array($comment->cid => $comment), $view_mode, $langcode);
```
In this and all subsequent `*_view()` wrappers, how would you feel about using `return $comment->view()` instead? Seems like more expected content for a wrapper, even though `$comment->view()` is ultimately going to call the same thing (so the current solution is less nesting and slightly more efficient).

herbdool replied:
I think it makes it a bit neater, sure.

herbdool replied:
@quicksketch I've made these changes and all tests are passing.

argiepiano commented:
I am doing further testing, and am encountering an unusual conflict with Entity Plus produced by the new method `buildContent`. I'm seeing this with the example module Basic Entity Plus Example, which uses Entity Plus, but this may cause issues for other ported modules that used D7's Entity API. 

At the center of this conflict is the fact that D7's Entity API's `buildContent` (defined in `EntityAPIController`)  returned a `$build` variable to the `view` method. This returned `$build` variable contained the renderable array. 

In Backdrop, the new `buildContent`  method doesn't return anything; rather, it stores the renderable array inside `$entity->content`, which is further processed by the `view` method.

In fact `buildContent` in D7 Entity API  actually **UNSET** `$entity->content` before returning `$build` (see [this line](https://git.drupalcode.org/project/entity/-/blob/7.x-1.x/includes/entity.controller.inc#L631))

So, the following potential conflicts (though rare) with custom entities are happening:
- A custom entity controller that relies on Backdrop's `buildContent` method, but define its own `view` method based on Entity API, will not receive a `$build` array from `buildContent`, OR:
- The opposite may happen too: an entity controller that defines its own `buildContent` based on Entity API, but relies on Backdrop's new `view` will have unset `$entity->content` before reaching Backdrop's `view`

As a possible solution, may I suggest that we modify `buildContent` to communicate in BOTH ways: by returning a `$build` render array (the old way), AND by leaving `$entity->content` intact, to accommodate the "new" way? Returning a `$build` array will certainly not cause any harm to core entities, which will simply ignore the returned array.

argiepiano commented:
Continuing some testing with other contrib modules to detect possible conflicts: 

The contrib module Registrations (maintained by @docwilmot and @hosef) uses a custom entity class called Registration that extends `Entity`. Its `view` method is incompatible with the new `Entity::view`, which produces a Fatal error when trying to display a Registration. 

This is a simple fix (the maintainers just need to add the missing arguments). But this illustrates the need to involve contrib maintainers testing these changes before they are released, and to coordinate the release with the release of new versions of these contrib projects.

I will continue checking other contrib modules (currently there are 22 projects in Github that depend on Entity Plus). <s>I checked Paragraphs (maintained by @laryn ) and that one seems to work fine with these changes. </s>

argiepiano commented:
I spoke too soon about Paragraphs. This one is experiencing the EXACT type of issue I mentioned in my comment above https://github.com/backdrop/backdrop/pull/3848#issuecomment-1002299809

The problem is that it uses `EntityPlusController::view`, but at the same time is using the new `Entity:buildContent` (which is invoked through `entity_plus_build_content()`). So,  `EntityPlusController::view` expects the call to `buildContent` to return a renderable array, but the new `Entity::buildContent` doesn't, which causes `Error: Unsupported operand types in EntityPlusController->view()`.

Pinging @laryn so that he's aware of this.

There are two possible solutions: either have `Entity::buildContent`  return a renderable array as I suggested in my comment above, or have `ParagraphsItemEntity` override `Entity::buildContent` with something as simple as this:

```php
 public function buildContent($view_mode = 'full', $langcode = NULL) {
    return entity_get_controller($this->entityType())->buildContent($this, $view_mode, $langcode); 
  }
```

(notice the `return` command - that's all that's needed).

herbdool commented:
@argiepiano I took a look and thought about it. I'm okay with having `buildContent` for the default entity and controller return the rendered array. I added it to the definition. Since we don't have to return the array for all entities, I've left it off for the non-default. There's no expectation currently anyway.

argiepiano commented:
> I took a look and thought about it. I'm okay with having `buildContent` for the default entity and controller return the rendered array. I added it to the definition

@herbdool , that's great. That solved the issues I mentioned above.

I've continued doing some LIGHT testing with the new code in this PR  (including your latest commit) with projects that depend on Entity Plus - specifically only those that define custom entities, and only those that have official releases. Here's a report:

- basic_entity_plus (maintained by me, an example module for Entity Plus): found no problem
- search_api (maintained by @earlyburg): found no problem
- membership_entity (there is no official release for this yet, maintained by @jayelless): found no problem
- rules (maintained by @laryn, @bugfolder and @Graham-72): found no problem   
- registration (maintained by @docwilmot and @hosef): found no problem (other than incompatible `Registration:view()` method I mentioned above)
- paragraphs (maintained by @laryn): found no problems after the latest commit in this PR
- ubercart (maintained by @laryn, @robertgarrigos and @bugfolder): I did find an issue that is not related to the problems with `buildContent` that the latest commit fixed. I have not been able to find what's causing it, but it's something that the maintainers would need to check prior to this PR being merged into core. 

Regarding the problem with Ubercart, since this PR is not merged yet, I decided not to create an issue in their queue, but I can briefly describe how to replicate it here, in case the maintainers want address it now: 
1. Create a Product 
2. Add product to cart
3. Visit the cart and click  "Checkout" (important: do not click "Update cart" before "Checkout")
4. The following notices appear when the checkout page is displayed: 
`Notice: Undefined index: total in theme_uc_cart_review_table() (line 530 of /Users/alejandrocremaschi/Sites/localhost/bdentity/modules/ubercart/uc_cart/uc_cart_checkout_pane.inc).`
The Cart Contents table doesn't show any product or prices. This doesn't happen if you test using a version of Backdrop without this PR.

I'm going to try running the automated tests for these modules above tomorrow and will report if there are any issues related to that. After that I think this will be RTBC

herbdool commented:
@argiepiano thank you for being thorough here. Quite helpful.

I'd like to note about the process here. After a pr is merged we still have a couple weeks to test and fix bugs/regressions before the release. So if this is just about a couple minor issues with contrib modules we'll still have time to get those fixes into the contrib modules or here.

herbdool commented:
@argiepiano the ubercart issue you found is weird. The method there doesn't even match entity_plus (see https://github.com/backdrop-contrib/ubercart/blob/54cd60913c411b0eff0eef759da7bbafa3890e48/uc_order/uc_order.controller.inc#L71), it's passing a `$content` variable back which doesn't appear in `entity_plus`. So it needs to get fixed there. Likely just adding the array to `$product->content` and returning it is enough.

bugfolder commented:
@argiepiano, many of the automated tests on Rules don't work yet (getting them to work is my current project). With the current PR on this issue https://github.com/backdrop-contrib/rules/issues/114, the checked-off tests should work, but RulesCoreIntegration still won't.

argiepiano commented:
> it's passing a `$content` variable back which doesn't appear in `entity_plus`. So it needs to get fixed there.

I agree that the ubercart maintainers need to look into this issue. The ubercart `buildContent` method does match the corresponding method implementation in Entity Plus - [see it here](https://github.com/backdrop-contrib/entity_plus/blob/4b5de214e6f4b8a58deb9f311a16cb796a879cf8/includes/entity_plus.controller.inc#L634). The `$content` variable was a way to pass content to the method - and in ubercart's case, to the parent. Still, I don't think the issue I found lies there.

argiepiano commented:
> @argiepiano, many of the automated tests on Rules don't work yet (getting them to work is my current project). With the current PR on this issue [backdrop-contrib/rules#114](https://github.com/backdrop-contrib/rules/issues/114), the checked-off tests should work, but RulesCoreIntegration still won't.

Yup, I noticed that.

quicksketch commented:
I have merged this into 1.x for 1.21.0. Thanks to @herbdool, @argiepiano, and @docwilmot for the feedback in this PR! Full credit list is in the commit and parent issue https://github.com/backdrop/backdrop-issues/issues/1739.
